bitbots_odometry/src/odometry_fuser.cpp

```diff
@@ -82,48 +102,126 @@ OdometryFuser::OdometryFuser() {
       tf2::Matrix3x3 imu_rotation_matrix(imu_orientation);
       imu_rotation_matrix.getRPY(imu_roll, imu_pitch, placeholder);
 
+      // get motion_odom transform
+      tf2::Transform motion_odometry;
+      tf2::fromMsg(_odom_data.pose.pose, motion_odometry);
+
       // get yaw from walking odometry
-      tf2::Quaternion odom_orientation;
-      tf2::fromMsg(_odom_data.pose.pose.orientation, odom_orientation);
+      tf2::Quaternion odom_orientation = motion_odometry.getRotation();
       tf2::Matrix3x3 odom_rotation_matrix(odom_orientation);
       odom_rotation_matrix.getRPY(placeholder, placeholder, walking_yaw);
 
       // combine orientations to new quaternion if IMU is active, use purely odom otherwise
-      tf2::Quaternion new_orientation;
+      tf2::Transform fused_odometry;
       if (imu_active) {
-        new_orientation.setRPY(imu_roll, imu_pitch, walking_yaw);
+        // compute the point of rotation (in base_link frame)
+        tf2::Transform rotation_point_in_base = getCurrentRotationPoint();
+        // publish rotation point as debug
+        geometry_msgs::TransformStamped rotation_point_msg;
+        rotation_point_msg.header.stamp = ros::Time::now();
+        rotation_point_msg.header.frame_id = "base_link";
+        rotation_point_msg.child_frame_id = "rotation";
+        geometry_msgs::Transform rotation_point_transform_msg;
+        rotation_point_transform_msg = tf2::toMsg(rotation_point_in_base);
+        rotation_point_msg.transform = rotation_point_transform_msg;
+        br.sendTransform(rotation_point_msg);
+        // get base_link in rotation point frame
+        tf2::Transform base_link_in_rotation_point = rotation_point_in_base.inverse();
+        // create transform that rotates around IMU roll,pitch and walk yaw
+        tf2::Quaternion rotation_quat;
+        rotation_quat.setRPY(imu_roll, imu_pitch, 0);
+        tf2::Transform rotation;
+        rotation.setRotation(rotation_quat);
+        rotation.setOrigin({0, 0, 0});
+
+        // transformation chain to get correctly rotated odom frame
+        // go to the rotation point in the odom frame. rotate the transform to the base link at this point
+        fused_odometry = motion_odometry * rotation_point_in_base * rotation * base_link_in_rotation_point;
```

Comment:
Correct me if I am wrong, but now the odom orientation and the IMU orientation (which is measured in an absolute scale) add up in the roll and pitch axis.

Reply:
I think you are right, I will correct it